pkg/systemd/notifyproxy/notifyproxy.go

```diff
@@ -110,48 +111,79 @@ func (p *NotifyProxy) WaitAndClose() error {
 		}
 	}()
 
-	const bufferSize = 1024
-	sBuilder := strings.Builder{}
-	for {
-		// Set a read deadline of one second such that we achieve a
-		// non-blocking read and can check if the container has already
-		// stopped running; in that case no READY message will be send
-		// and we're done.
-		if err := p.connection.SetReadDeadline(time.Now().Add(time.Second)); err != nil {
-			return err
-		}
-
+	// Since reading from the connection is blocking, we need to spin up two
+	// goroutines.  One waiting for the `READY` message, the other waiting
+	// for the container to stop running.
+	errorChan := make(chan error, 1)
+	readyChan := make(chan bool)
```

Comment:
I think this needs a defer close for both channels inside their goroutines to prevent write after close problems?

Reply:
Wouldn't that cause a write after close?  Let's assume we have a read error and the container exits.  If routine 1 closes the channel that routine 2 wants to write the error to, we'd run into this issue.

Reply:
I made both channels buffered to make sure none of the two routines can block.  Once both are done, the channels will get garbage collected.  But there is no guarantee that both routines will be finished once the function returns.  There is a chance that the container sends ready, does something and exits. Routine 1 receives the READY, routine 2 detects the container isn't running and sends the error before `ctx` has been cancelled.

Reply:
I was trying to say close them inside goroutine with defer, i.e. on line 121 and 160

Reply:
Yes, I understood.  You owe an explanation though.  Only when closing the channels, they are subject to a write after close.  Since they're buffered, the routines cannot block and the channel will be garbage collected when both routines returned. I think that closing makes it more complicated than necessary. But I may be missing something.